### PR TITLE
Org mgt 3

### PR DESCRIPTION
**Pull request by andrea-grapl**
### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/667

### What changes does this PR make to Grapl? Why?
https://github.com/grapl-security/grapl-rfcs/tree/org-and-user-management

### How were these changes tested?
`src/rust/organization-management/tests`

### CODE REVIEW

File: nomad/grapl-core.nomad [hunk 1]
```diff
@@ -123,6 +124,26 @@ variable "plugin_registry_kernel_artifact_url" {
...
+
+variable "organization_management_db_username" {
+  type        = string
+  description = "What is the username for the organization management table?"
+}
+
+variable "organization_management_db_password" {
+  type        = string
+  description = "What is the password for the organization management table?"
+}
```
Comment (christophermaier) [inline nomad/grapl-core.nomad:145]: These descriptions should be tweaked to refer to a "database" rather than a "table".

File: nomad/local/grapl-local-infra.nomad [hunk 2]
```diff
@@ -65,6 +65,18 @@ variable "PLUGIN_WORK_QUEUE_DB_PASSWORD" {
   default     = "postgres"
 }
 
+variable "ORGANIZATION_MANAGEMENT_DB_USERNAME" {
```
Comment (christophermaier) [inline nomad/local/grapl-local-infra.nomad:68]: Not really a comment about this PR _per se_, but is there an important distinction between all-uppercase variables and all-lowercase variables? I just noticed these differ from the ones in `grapl-core.nomad` and `integration-tests.nomad` in this respect.

File: pulumi/grapl/__main__.py [hunk 3]
```diff
@@ -310,6 +311,10 @@ def main() -> None:
         ###################################
         # Local Grapl
         ###################################
+        organization_management_db = LocalPostgresInstance(
+            name="organization_management_db",
+            port=5632,
```
Comment (christophermaier) [inline pulumi/grapl/__main__.py:316]: Is there a reason this is defined both here and in `local-grapl.env`? It would be better to have a single source of truth.

**Review by christophermaier — COMMENTED** [review]
I'm not done reviewing, but have to step away for a bit, so I'm publishing my comments thus far.

File: src/rust/organization-management/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+- Table definitions can be found and altered in the migration file in the
+  `migrations` folder
+- We can apply table changes, generate the `sql-data.json` file by running our
+  rebuild script `./rebuild.sh` in the `organization-management` folder. This
+  script creates and populates a local database in a docker container.
+
+### The path to the proto files are:
+
+`./src/proto/graplinc/grapl/api/organization_management/v1beta1/organization_management.proto`
```
Comment (christophermaier) [inline src/rust/organization-management/README.md:13]: It would be better to leave the leading `./` off of this path; otherwise, it implies that this is found within this crate.

File: src/rust/organization-management/README.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+## Organization Management
+
+### When making changes to the organization management service:
+
+- Table definitions can be found and altered in the migration file in the
+  `migrations` folder
```
Comment (christophermaier) [inline src/rust/organization-management/README.md:6]: I'd reword this a bit; as it is, it implies there is a single file to interact with, and that developers should make changes to it, rather than creating new migrations.

It would also be good to say what the command to generate a new migration file is.

File: src/rust/organization-management/README.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  rebuild script `./rebuild.sh` in the `organization-management` folder. This
+  script creates and populates a local database in a docker container.
+
+### The path to the proto files are:
+
+`./src/proto/graplinc/grapl/api/organization_management/v1beta1/organization_management.proto`
+
+### Changes to the above `.proto` file should also be applied to:
+
+`src/rust/rust-proto/src/organization_management.rs`
```
Comment (christophermaier) [inline src/rust/organization-management/README.md:17]: One other usability win here would be to make all the files mentioned in this document into hyperlinks; see [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-readmes#relative-links-and-image-paths-in-readme-files) for details.

File: src/rust/organization-management/rebuild-sqlx.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
```
Comment (christophermaier) [inline src/rust/organization-management/rebuild-sqlx.sh:1]: You'll want to add a `BUILD` file to `src/rust/organization-management` with the following contents:

```
shell_sources()
```
This will ensure that the shell scripts in this directory are formatted and linted (and even tested, if you like!) by Pants.

File: src/rust/organization-management/rebuild-sqlx.sh [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    --detach \
+    --publish 5432:5432 \
+    --rm \
+    --name ${CONTAINER_NAME} \
+    --env POSTGRES_PASSWORD \
+    postgres
+
+trap cleanup EXIT INT
+
+echo ""
```
Comment (christophermaier) [inline src/rust/organization-management/rebuild-sqlx.sh:26]: What was this `echo` for? It looks like it can just be deleted, unless you want to have a message like `"Sleeping for 5 seconds to allow the database to come up"`.

File: src/rust/organization-management/rebuild-sqlx.sh [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+export DATABASE_URL=postgres://${USER}:${POSTGRES_PASSWORD}@localhost/${DBNAME}
+
+echo "Creating PostGresDB"
+sqlx database create
+
+echo "Running migrations"
+sqlx migrate run
+
+#cargo build
```
Comment (christophermaier) [inline src/rust/organization-management/rebuild-sqlx.sh:37]: This could be deleted.

File: src/rust/organization-management/rebuild-sqlx.sh [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+echo "Saving metadata to sqlx.json for offline mode"
+cargo sqlx prepare -- --lib   
+
+#cargo run --bin organization-management
```
Comment (christophermaier) [inline src/rust/organization-management/rebuild-sqlx.sh:42]: This could also be deleted.

File: src/rust/organization-management/setup.sh [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+echo "Running migrations"
+sqlx migrate run
+
+cargo build
+
+echo "Saving metadata to sqlx.json for offline mode"
+cargo sqlx prepare -- --lib
+
+echo "Running org management"
+cargo run
```
Comment (christophermaier) [inline src/rust/organization-management/setup.sh:43]: This script is identical to the `rebuild-sqlx.sh` script with the exception of the `cargo build` and `cargo run` lines. If both scripts are truly necessary, it would be better to consolidate them into a single script, and modify its behavior based on an argument.

(That would also be a good first step toward an ultimate generalization of this script so it can be used when working on _any_ of our SQLX-using services. _That_ would be something for a future PR, though, but seems like a natural evolution of what you have here.)

File: src/rust/organization-management/src/client.rs [hunk 12]
```diff
@@ -0,0 +1,98 @@
+use std::{
+    convert::TryInto,
+    time::Duration,
+};
+
+use rust_proto::organization_management::{
+    organization_management_service_client::OrganizationManagementServiceClient as _OrganizationManagementServiceClient,
```
Comment (christophermaier) [inline src/rust/organization-management/src/client.rs:7]: In other GRPC services, I believe we settled on a convention of naming things like this `FooProto`, rather than `_Foo`, because the leading underscore implies something that is not used.

File: src/rust/organization-management/src/client.rs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        &mut self,
+        request: CreateUserRequest,
+    ) -> Result<CreateUserResponse, OrganizationManagementServiceClientError> {
+        let response = self
+            .inner
+            .create_user(CreateUserRequestProto::from(request))
+            .await?;
+        Ok(response.into_inner().try_into()?)
+    }
+}
```
Comment (christophermaier) [inline src/rust/organization-management/src/client.rs:98]: I can't comment on the file directly because it's empty, but what is supposed to be in `src/rust/organization-management/src/db/client.rs`?

File: src/rust/organization-management/src/server.rs [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        let response: CreateUserResponseProto = response.into();
+        Ok(Response::new(response))
+    }
+}
+
+#[tracing::instrument(err)]
+async fn create_db_connection() -> Result<Pool<Postgres>, sqlx::Error> {
+    let url = "postgres://postgres@localhost?db_name=postgres&user=postgres&password=postgres";
+
+    println!("databse url {}", url);
```
Comment (christophermaier) [inline src/rust/organization-management/src/server.rs:257]: We shouldn't have any `println` calls; we're using the tracing infrastructure for this kind of stuff.

File: src/rust/organization-management/src/server.rs [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+
+        let response = self.create_user(request).await?;
+        let response: CreateUserResponseProto = response.into();
+        Ok(Response::new(response))
+    }
+}
+
+#[tracing::instrument(err)]
+async fn create_db_connection() -> Result<Pool<Postgres>, sqlx::Error> {
+    let url = "postgres://postgres@localhost?db_name=postgres&user=postgres&password=postgres";
```
Comment (christophermaier) [inline src/rust/organization-management/src/server.rs:255]: We shouldn't be hardcoding the connection URL; it should be taking whatever configuration values we've passed the server.

File: src/rust/organization-management/src/server.rs [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+
+    tracing::info!(message="connecting to postgres", url=%url);
+
+    // Create Connection Pool
+    let pool = PgPoolOptions::new()
+        .max_connections(5)
+        .connect(&url)
+        .await?;
+
+    // Insert Org Info
```
Comment (christophermaier) [inline src/rust/organization-management/src/server.rs:267]: This comment seems out of place here.

File: src/rust/plugin-registry/tests/test_create_plugin.rs [hunk 17]
```diff
@@ -21,6 +21,7 @@ async fn test_create_plugin() -> Result<(), Box<dyn std::error::Error>> {
     let tenant_id = uuid::Uuid::new_v4();
```
Comment (christophermaier) [inline src/rust/plugin-registry/tests/test_create_plugin.rs:21]: This file shouldn't be in this PR, since it isn't related to organization management.

File: src/rust/rust-proto/Cargo.toml [hunk 18]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "rust-proto"
 version = "0.2.10"
-authors = ["Insanitybit <insanitybit@gmail.com>"]
+authors = ["Grapl <colin@graplsecurity.com>"]
```
Comment (christophermaier) [inline src/rust/rust-proto/Cargo.toml:4]: This is also a weird change to have in this PR.

Comment (andrea-grapl) [reply src/rust/rust-proto/Cargo.toml:4]: reverted

File: nomad/local/grapl-local-infra.nomad (hunk 2 above)
Comment (andrea-grapl) [reply nomad/local/grapl-local-infra.nomad:68]: Not that I'm aware of, I tried to follow existing implementations which used all uppercase, but we can change this if needed. Leaving for now.

File: nomad/local/grapl-local-infra.nomad [hunk 19]
```diff
@@ -442,4 +460,47 @@ job "grapl-local-infra" {
...
+      }
+
+      service {
+        name = "organization-management-db"
+
+        check {
+          type     = "script"
+          name     = "check_postgres"
+          command  = "pg_isready"
+          args     = ["-U", "postgres"]
```
[the remark this answers is not in the document]
Comment (andrea-grapl) [reply nomad/local/grapl-local-infra.nomad:493]: Leaving for now, I modeled the `organization-management` group after other groups using postgres like `plugin-work-queue` and `plugin-registry`. I'm happy to change this but will do so in a separate PR and have created an issue here: https://github.com/grapl-security/issue-tracker/issues/872

File: src/rust/rust-proto/Cargo.toml (hunk 18 above)
Comment (andrea-grapl) [reply src/rust/rust-proto/Cargo.toml:4]: Reverted

File: src/rust/organization-management/README.md (hunk 6 above)
Comment (andrea-grapl) [reply src/rust/organization-management/README.md]: thanks

File: nomad/local/grapl-local-infra.nomad (hunk 2 above)
Comment (andrea-grapl) [reply nomad/local/grapl-local-infra.nomad:68]: I'm not particularly sure, I followed standards set by other services, this is a good thing to consider though. Leaving for now but would be happy to refactor this service along with others to have consistent casing if that makes more sense. Leaving for now.

File: nomad/local/integration-tests.nomad [hunk 20]
```diff
@@ -94,6 +94,28 @@ variable "plugin_work_queue_db_password" {
...
 
+
+variable "organization_management_db_hostname" {
+  type        = string
+  description = "The host for the local organization management  db"
+}
+
+variable "organization_management_db_port" {
+  type        = string
+  default     = "5432"
```
[the remark this answers is not in the document]
Comment (andrea-grapl) [reply nomad/local/integration-tests.nomad]: Removed

File: pulumi/grapl/__main__.py (hunk 3 above)
Comment (andrea-grapl) [reply pulumi/grapl/__main__.py:316]: Not sure, @colin-grapl  - any input?^^

File: nomad/grapl-core.nomad [hunk 21]
```diff
@@ -12,6 +12,7 @@ variable "container_images" {
   type        = map(string)
   description = <<EOF
   A map of $NAME_OF_TASK to the URL for that task's docker image ID.
+  A map of $NAME_OF_TASK to the URL for that task's docker image ID.
```
Comment (colin-grapl) [inline nomad/grapl-core.nomad:15]: Is this a duplicate line?

File: pulumi/grapl/__main__.py (hunk 3 above)
Comment (colin-grapl) [reply pulumi/grapl/__main__.py:316]: I have no idea how all of these things work together. I don't know how you would access the env here.

File: src/rust/organization-management/setup.sh (hunk 11 above)
Comment (colin-grapl) [reply src/rust/organization-management/setup.sh:43]: Remove setup.sh

File: src/rust/organization-management/src/client.rs (hunk 13 above)
Comment (colin-grapl) [reply src/rust/organization-management/src/client.rs:98]: Looks like it was probably intended to be a wrapper for the sqldb itself. Kill it imo.

File: src/rust/organization-management/src/server.rs (hunk 15 above)
Comment (colin-grapl) [reply src/rust/organization-management/src/server.rs:255]: Is this dead code? I don't see this called anywhere.

Comment (andrea-grapl) [reply src/rust/organization-management/src/server.rs:255]: yes, dead code

**Review by colin-grapl — APPROVED** [review]
